File: schedule/video.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
@Author: reber
@Mail: reber0ask@qq.com
@Date: 2021-01-06 09:44:54
LastEditTime: 2021-11-28 12:11:10
'''

import re
import time
import demjson
from lxml import etree
from sqlalchemy import func

from sqlmodule import session_maker
from sqlmodule import to_dict
from sqlmodule import User
from sqlmodule import Video
from sqlmodule import Data
from sqlmodule import Message

from lib.data import global_data
from lib.request import req, ReqException


def video_job():
    global_data.logger.info("==> video check...")

    logger_msg(msg_type="system", username="schedule", action="start video check")
    _video = VideoClass()
    _video.check()

def logger_msg(username="-", action="-", data="-", msg_type=""):
    """
    记录后天计划任务、程序错误信息等
    """
    with session_maker(global_data.sqlite_uri) as db_session:
        msg = Message(username=username, action=action, data=data, status="unread", msg_type=msg_type)
        db_session.add(msg)

class VideoClass(object):
    def __init__(self):
        super(VideoClass, self).__init__()

    def get_all_video_msg(self):
        """
        获取未完结的站点的信息
        """
        video_msg_list = list()
        try:
            with session_maker(global_data.sqlite_uri) as db_session:
                results = db_session.query(Video).filter(Video.status != "已完结").all()
                video_msg_list = to_dict(results)
        except Exception as e:
            global_data.logger.error(str(e))
            return []
        else:
            # print(video_msg_list)
            return video_msg_list

    def get_video_exist_url(self, site_id):
        """
        获取一个动漫的现有链接
        """
        video_msg_list = list()
        video_url_list = list()
        try:
            with session_maker(global_data.sqlite_uri) as db_session:
                results = db_session.query(Data).filter_by(ref_id=site_id).filter_by(category='video').all()
                video_msg_list = to_dict(results)
        except Exception as e:
            global_data.logger.error(str(e))
        else:
            if video_msg_list:
                video_url_list = [video_msg["url"] for video_msg in video_msg_list]
        return video_url_list

    def get_new_video_msg(self, username, name, link, video_url_list):
        """
        获取一个动漫的新剧集的信息
        """
        href_text_list = list()
        status = ""
        video_type = ""

        if link.startswith("https://space.bilibili.com/"):
            href_text_list, status, video_type = self.bilibili_up(username, name, link)
        elif link.startswith("https://www.bilibili.com/bangumi"):
            href_text_list, status, video_type = self.bilibili_bangumi(username, name, link)
        elif link.startswith("https://www.acfun.cn/u"):
            href_text_list, status, video_type = self.acfun_up(username, name, link)
        elif link.startswith("https://www.acfun.cn/bangumi"):
            href_text_list, status, video_type = self.acfun_bangumi(username, name, link)
        elif link.startswith("https://www.agefans.vip/"):
            href_text_list, status, video_type = self.age(username, name, link)
        elif link.startswith("http://www.yhdm2.com/"):
            href_text_list, status, video_type = self.yhdm2(username, name, link)

        new_video_msg_list = list()
        for href_text in href_text_list:
            url, title = href_text
            # html = "<iframe src={}></iframe>".format(url)
            html = "<h2>双指向左滑动</h2>"
            if url not in video_url_list:
                new_video_msg_list.append((title, url, html))

        return new_video_msg_list, status, video_type

    def check(self):
        video_msg_list = self.get_all_video_msg()
        for video_msg in video_msg_list:
            site_id = video_msg.get("id")
            user_id = video_msg.get("user_id")
            name = video_msg.get("name")
            link = video_msg.get("link")
            status = video_msg.get("status")

            with session_maker(global_data.sqlite_uri) as db_session:
                result = db_session.query(User.username).filter_by(user_id=user_id).first()
                username = result.username

            # global_data.logger.info("Video check: {}".format(link))
            video_url_list = self.get_video_exist_url(site_id)
            new_video_msg_list, curr_status, video_type = self.get_new_video_msg(username, name, link, video_url_list)

            article_data = list()
            for new_video_msg in new_video_msg_list:
                title, url, html = new_video_msg
                date = func.now()

                article = Data(ref_id=site_id, category="video", title=title, url=url, date=date, description=html, status="unread")
                article_data.append(article)

            if article_data:
                message_data_list = [x[0] for x in new_video_msg_list] # 获取 title 的 list 用于保存
                logger_msg(msg_type="system", username="schedule", action="video check: {}".format(name), data=str(message_data_list))
                logger_msg(msg_type="user", username=username, action="{} 已更新".format(name), data=str(message_data_list))
                
                affect_num = 0
                try:
                    with session_maker(global_data.sqlite_uri) as db_session:
                        db_session.add_all(article_data)
                        if curr_status != status:
                            affect_num = db_session.query(Video).filter_by(id=site_id).update({"status": curr_status})
                            if affect_num:
                                global_data.logger.info("check {} success".format(name))
                except Exception as e:
                    global_data.logger.error(str(e))
                else:
                    if affect_num:
                        global_data.logger.info("Video check: {} 获取 {} 个新资源".format(name, len(article_data)))

    def bilibili_up(self, username, name, url):
        href_text_list = list()
        status = ""
        vedio_type = "bilibili_up"

        # up主视频接口
        up_video_api = "https://api.bilibili.com/x/space/arc/search?mid={}&ps=30&tid=0&pn=1&keyword=&order=pubdate&jsonp=jsonp"

        # 获取视频列表
        up_id = url.strip("/").split("/")[-1]
        try:
            resp = req.get(up_video_api.format(up_id))
            result = resp.json()
        except ReqException as e:
            global_data.logger.error(e)
        else:
            if result.get("code") == 0:
                vlist = result.get("data").get("list").get("vlist")
                for v in vlist:
                    title = v.get("title")
                    bvid = v.get("bvid")
                    url = "https://www.bilibili.com/video/{}".format(bvid)
                    href_text_list.append((url, title))

            href_text_list = href_text_list[::-1]
            status = "连载中"

        return href_text_list, status, vedio_type

    def bilibili_bangumi(self, username, name, url):
        href_text_list = list()
        status = ""
        vedio_type = "bilibili_bangumi"

        #番剧接口
        bangumi_api = "https://api.bilibili.com/pgc/view/web/season?season_id={}"

        # 获取视频列表
        try:
            resp = req.get(url)
            m = re.search(r'season_id":(\d+),', resp.text, re.M|re.S)
            season_id = m.group(1)
        except ReqException as e:
            global_data.logger.error(e)
        except Exception as e:
            # 番剧下架的话会获取不到 season_id
            # 出现 AttributeError: 'NoneType' object has no attribute 'group'
            if "NoneType' object has no attribute 'group" in str(e):
                msg = "未能成功获取 season_id，可能下架了..."
                global_data.logger.error(name+msg)
                logger_msg(msg_type="system", username="schedule", action="video check: {}".format(name), data=msg)
                logger_msg(msg_type="user", username=username, action="video check: {}".format(name), data=msg)
        else:
            resp = req.get(bangumi_api.format(season_id))
            result = resp.json()
            if result.get("code") == 0:
                result = result.get("result")
                _title = result.get("title")

                # 获取正片
                episodes = result.get("episodes")
                for episode in episodes:
                    badge = episode.get("badge") # 值为 预告 或为空
                    if "预告" not in badge:
                        title = episode.get("share_copy").replace("《" + _title + "》", "").strip()
                        url = episode.get("share_url")
                        href_text_list.append((url, title))

                status = result.get("publish").get("is_finish")
                status = "已完结" if status else "连载中"

        return href_text_list, status, vedio_type

    def acfun_up(self, username, name, url):
        href_text_list = list()
        status = ""
        vedio_type = "acfun_up"

        try:
            html = req.get(url).text
        except ReqException as e:
            global_data.logger.error(e)
        except Exception as e:
            global_data.logger.error(str(e))
        else:
            a_href_list = re.findall(r'<a href="(.*?)" target="_blank" class="ac-space-video', html)
            movurl_list = ["https://www.acfun.cn"+a_href for a_href in a_href_list]
            title_list = re.findall(r'<figcaption>.*?<p class="title line" title=".*?">(.*?)</p>', html, re.S|re.M)

            for index,title in enumerate(title_list):
                url = movurl_list[index]
                href_text_list.append((url, title))

            href_text_list = href_text_list[::-1]
            status = "连载中"

        return href_text_list, status, vedio_type

    def acfun_bangumi(self, username, name, url):
        href_text_list = list()
        status = ""
        vedio_type = "acfun_bangumi"

        try:
            html = req.get(url).text

            status = re.search(r'extendsStatus":"(.*?)",', html, re.S|re.M).group(1)
            bangumiList = re.search(r'window\.bangumiList = (.*?);', html).group(1)
        except ReqException as e:
            global_data.logger.error(e)
        except Exception as e:
            global_data.logger.error(str(e))
        else:
            items = demjson.decode(bangumiList).get("items")
            items = sorted(items, key=lambda x:x["priority"], reverse=True)
            for item in items:
                title = item.get("title")
                episodeName = item.get("episodeName")
                title = "{} {}".format(title, episodeName)

                bangumiId = item.get("bangumiId")
                itemId = item.get("itemId")
                priority = int(item.get("priority"))
                if priority != 1000:
                    tmp_id = "{:1<6d}".format(priority)
                    url = "https://www.acfun.cn/bangumi/aa{}_{}_{}".format(bangumiId, tmp_id, itemId)
                    href_text_list.append((url, title))

            href_text_list = href_text_list[::-1]

        return href_text_list, status, vedio_type

    def age(self, username, name, url):
        href_text_list = list()
        status = "连载中"
        vedio_type = "age"

        try:
            resp = req.get(url)
            resp.encoding = resp.apparent_encoding
            html = resp.text
        except ReqExceptin as error_msg:
            global_data.logger.error(error_msg)
            logger_msg(msg_type="system", username="schedule", action="video check: {}".format(name), data=str(error_msg))
            logger_msg(msg_type="user", username=username, action="{} 更新".format(name), data=str(error_msg))
        else:
            selector = etree.HTML(html)
            li_tag_list = selector.xpath('//*/div[@style="display:block"]/ul/li')
            for li_tag in li_tag_list:
                url = "https://www.agefans.vip"+li_tag.xpath('a/@href')[0]
                title = li_tag.xpath('a/text()')[0]
                if "pv" in title.lower() or "无字" in title:
                    continue
                href_text_list.append((url, title))

            detail_imform_kv_list = selector.xpath('//*/li[@class="detail_imform_kv"]')
            status_kv = detail_imform_kv_list[7]
            status = status_kv.xpath('span[2]/text()')[0]
            if status == "完结":
                status = "已完结"
            elif status == "连载":
                status = "连载中"

        return href_text_list, status, vedio_type

    def yhdm2(self, username, name, url):
        href_text_list = list()
        status = "连载中"
        vedio_type = "yhdm2"

        try:
            resp = req.get(url)
            resp.encoding = resp.apparent_encoding
            html = resp.text
        except ReqExceptin as error_msg:
            global_data.logger.error(error_msg)
            logger_msg(msg_type="system", username="schedule", action="video check: {}".format(name), data=str(error_msg))
            logger_msg(msg_type="user", username=username, action="{} 更新".format(name), data=str(error_msg))
        else:
            selector = etree.HTML(html)
            li_tag_list = selector.xpath('//*[@id="stab_1_71"]/ul/li')
            for li_tag in li_tag_list:
                url = "http://www.yhdm2.com"+li_tag.xpath('a/@href')[0]
                title = li_tag.xpath('a/text()')[0]
                if "pv" in title.lower() or "备用" in title or "英字" in title or "生肉" in title:
                    continue
                href_text_list.append((url, title))

            if status != "已完结":
                href_text_list = href_text_list[::-1]

        return href_text_list, status, vedio_type

File: schedule/video.go
/*
 * @Author: reber
 * @Mail: reber0ask@qq.com
 * @Date: 2022-01-04 21:12:34
 * @LastEditTime: 2024-01-24 13:52:50
 */
package schedule

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/bitly/go-simplejson"
	"github.com/reber0/goutils"
	"github.com/reber0/rss-sub/global"
	"github.com/reber0/rss-sub/mydb"
	"github.com/tidwall/gjson"
)

func checkVideo() {
	global.Log.Info("checkVideo...")

	videoAllSiteMsg := getVideoAllSiteMsg()
	for _, siteMsg := range videoAllSiteMsg {
		site_id := siteMsg.ID
		user_id := siteMsg.UID
		name := siteMsg.Name
		link := siteMsg.Link

		var username string
		result := global.Db.Model(&mydb.User{}).Select("uname").Where("uid=?", user_id).First(&username)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
		}

		global.Log.Info(fmt.Sprintf("schedule video check ==> %s", name))
		videoURLSlice := getVideoURL(site_id)
		newVideoMsgList, status, err := getNewVideoMsg(link, videoURLSlice)
		if err != nil {
			loggerMsg(username, name+" 更新", err.Error())
		} else {
			if len(newVideoMsgList) > 0 {
				if len(newVideoMsgList) > 30 {
					newVideoMsgList = newVideoMsgList[0:30]
				}

				// 保存新的番剧信息
				var titleSlice []string
				for _, newVideoMsg := range newVideoMsgList {
					title := newVideoMsg[0]
					url := newVideoMsg[1]
					data := mydb.Data{RefId: site_id, Category: "video", Title: title, URL: url, Status: "unread"}
					global.Db.Create(&data)

					titleSlice = append(titleSlice, title)
				}
				loggerMsg(username, name+" 更新", strings.Join(titleSlice, " "))

				// 更新番剧状态
				result := global.Db.Model(&mydb.Video{}).Where("id = ?", site_id).Update("status", status)
				if result.Error != nil {
					global.Log.Error(result.Error.Error())

				}
			}
		}
	}
}

// 获取未完结的站点的信息
func getVideoAllSiteMsg() []mydb.Video {
	var siteMsgSlice []mydb.Video
	result := global.Db.Model(&mydb.Video{}).Where("status != '已完结'").Find(&siteMsgSlice)
	if result.Error != nil {
		global.Log.Error(result.Error.Error())
	}
	return siteMsgSlice
}

// 获取一个番剧已获取的链接
func getVideoURL(videoID int) []string {
	var videoURLSlice []string

	var dataSlice []mydb.Data
	result := global.Db.Model(&mydb.Data{}).Where("category='video' and ref_id = ?", videoID).Find(&dataSlice)
	if result.Error != nil {
		global.Log.Error(result.Error.Error())
	} else {
		for _, data := range dataSlice {
			videoURLSlice = append(videoURLSlice, data.URL)
		}
	}

	return videoURLSlice
}

// 获取一个番剧更新的信息
func getNewVideoMsg(targetURL string, videoURLSlice []string) ([][]string, string, error) {
	var status string
	var newVideoMsgList [][]string

	baseURL := goutils.NewURL(targetURL).BaseURL()

	var err error
	var href_text_list [][]string
	if strings.Contains(targetURL, "space.bilibili.com") {
		href_text_list, status, err = bilibiliUp(targetURL)
	} else if strings.Contains(targetURL, "www.bilibili.com/bangumi") {
		href_text_list, status, err = bilibiliBangumi(targetURL)
	} else if strings.Contains(targetURL, "acfun.cn/u") {
		href_text_list, status, err = acfunUp(targetURL)
	} else if strings.Contains(targetURL, "acfun.cn/bangumi") {
		href_text_list, status, err = acfunBangumi(targetURL)
	} else if strings.Contains(targetURL, "ysjdm.net") {
		href_text_list, status, err = ysjdm(targetURL)
	} else if strings.Contains(targetURL, "www.yinghuacd.com") {
		href_text_list, status, err = yinghuacd(targetURL)
	} else if strings.Contains(targetURL, "age") {
		href_text_list, status, err = age(targetURL)
	}
	if err != nil {
		return newVideoMsgList, status, err
	} else {
		for _, href_text := range href_text_list {
			a_text := strings.TrimSpace(href_text[0])

			a_href := strings.TrimSpace(href_text[1])
			if !strings.HasPrefix(a_href, "http") {
				a_href = baseURL + strings.TrimLeft(a_href, "/")
			}

			// 暂存新剧集
			if !goutils.IsInCol(a_href, videoURLSlice) {
				newVideoMsgList = append(newVideoMsgList, []string{a_text, a_href})
			}
		}
	}

	return newVideoMsgList, status, nil
}

func bilibiliUp(targetURL string) ([][]string, string, error) {
	var newVideoMsgList [][]string

	// up主视频接口
	up_video_api := "https://api.bilibili.com/x/space/arc/search?mid={mid}&ps=30&tid=0&pn=1&keyword=&order=pubdate&jsonp=jsonp"

	// 获取视频列表
	_tmp := strings.Split(strings.Trim(targetURL, "/"), "/")
	up_id := _tmp[len(_tmp)-1]
	resp, err := global.Client.R().Get(strings.ReplaceAll(up_video_api, "{mid}", up_id))
	if err != nil {
		global.Log.Error(err.Error())
		return newVideoMsgList, "连载中", err
	} else {
		html := goutils.EncodeToUTF8(resp)

		code := gjson.Get(html, "code").Int()
		if code == 0 {
			vlist := gjson.Get(html, "data.list.vlist").Array()
			for _, v := range vlist {
				title := v.Get("title").String()
				bvid := v.Get("bvid").String()
				url := strings.ReplaceAll("https://www.bilibili.com/video/{bvid}", "{bvid}", bvid)
				newVideoMsgList = append(newVideoMsgList, []string{title, url})
			}
			newVideoMsgList = goutils.SliceListReverse(newVideoMsgList)
		}
	}

	return newVideoMsgList, "连载中", nil
}

func bilibiliBangumi(targetURL string) ([][]string, string, error) {
	var status string
	var newVideoMsgList [][]string

	// 番剧接口
	bangumi_api := "https://api.bilibili.com/pgc/view/web/season?season_id={season_id}"

	// 获取视频列表
	resp, err := global.Client.R().Get(targetURL)
	if err != nil {
		global.Log.Error(err.Error())
		return newVideoMsgList, "连载中", err
	} else {
		html := goutils.EncodeToUTF8(resp)

		reg := regexp.MustCompile(`season_id":(\d+),`)
		m := reg.FindStringSubmatch(html)
		season_id := m[1]

		resp, err := global.Client.R().Get(strings.ReplaceAll(bangumi_api, "{season_id}", season_id))
		if err != nil {
			global.Log.Error(err.Error())
			return newVideoMsgList, "连载中", err
		} else {
			html := goutils.EncodeToUTF8(resp)

			code := gjson.Get(html, "code").Int()
			if code == 0 {
				// result := gjson.Get(html, "result")
				_title := gjson.Get(html, "result.title").String()

				// 获取正片
				episodes := gjson.Get(html, "result.episodes").Array()
				for _, episode := range episodes {
					badge := episode.Get("badge").String()
					if !strings.Contains(badge, "预告") {
						share_copy := episode.Get("share_copy").String()
						title := strings.TrimSpace(strings.ReplaceAll(share_copy, "《"+_title+"》", ""))
						url := episode.Get("share_url").String()
						newVideoMsgList = append(newVideoMsgList, []string{title, url})
					}
				}
				if gjson.Get(html, "result.publish.is_finish").Int() == 1 {
					status = "已完结"
				} else {
					status = "连载中"
				}
			}
		}

	}

	return newVideoMsgList, status, nil
}

func acfunUp(targetURL string) ([][]string, string, error) {
	var newVideoMsgList [][]string

	resp, err := global.Client.R().Get(targetURL)
	if err != nil {
		global.Log.Error(err.Error())
		return newVideoMsgList, "连载中", err
	} else {
		html := goutils.EncodeToUTF8(resp)

		var movurl_list []string
		reg1 := regexp.MustCompile(`<a href="(.*?)" target="_blank" class="ac-space-video`)
		for _, a_href := range reg1.FindAllStringSubmatch(html, -1) {
			movurl_list = append(movurl_list, a_href[1])
		}

		reg2 := regexp.MustCompile(`(?sm)<figcaption>.*?<p class="title line" title="(.*?)">`)
		for index, title := range reg2.FindAllStringSubmatch(html, -1) {
			url := movurl_list[index]
			newVideoMsgList = append(newVideoMsgList, []string{title[1], url})
		}
		newVideoMsgList = goutils.SliceListReverse(newVideoMsgList)
	}

	return newVideoMsgList, "连载中", nil
}

func acfunBangumi(targetURL string) ([][]string, string, error) {
	var status string
	var newVideoMsgList [][]string

	resp, err := global.Client.R().Get(targetURL)
	if err != nil {
		global.Log.Error(err.Error())
		return newVideoMsgList, "连载中", err
	} else {
		html := goutils.EncodeToUTF8(resp)

		reg1 := regexp.MustCompile(`(?sm)extendsStatus":"(.*?)",`)
		m1 := reg1.FindStringSubmatch(html)
		status = m1[1]

		reg2 := regexp.MustCompile(`window\.bangumiList = (.*?);`)
		m2 := reg2.FindStringSubmatch(html)
		bangumiList := m2[1]

		res, err := simplejson.NewJson([]byte(bangumiList))
		if err != nil {
			global.Log.Error(err.Error())
		}
		for _, item := range res.Get("items").MustArray() {
			item := item.(map[string]interface{})

			title := fmt.Sprintf("%s %s", item["title"], item["episodeName"])

			bangumiId := item["bangumiId"]
			priority := item["priority"]
			itemId := item["itemId"]
			if priority != "1000" {
				url := fmt.Sprintf("https://www.acfun.cn/bangumi/aa%s_%s_%s", bangumiId, priority, itemId)
				newVideoMsgList = append(newVideoMsgList, []string{title, url})
			}
		}
	}

	return newVideoMsgList, status, nil
}

func ysjdm(targetURL string) ([][]string, string, error) {
	var status string
	var newVideoMsgList [][]string

	resp, err := global.Client.R().Get(targetURL)
	if err != nil {
		global.Log.Error(err.Error())
		return newVideoMsgList, "连载中", err
	} else {
		html := goutils.EncodeToUTF8(resp)
		dom, err := goquery.NewDocumentFromReader(strings.NewReader(html))
		if err != nil {
			global.Log.Error(err.Error())
		}

		dom.Find(`ul[class="content_playlist clearfix"]`).Eq(0).Find(`li>a`).Each(
			func(i int, node *goquery.Selection) {
				url, _ := node.Attr("href")
				title := node.Text()
				if !strings.Contains(strings.ToLower(title), "pv") && !strings.Contains(strings.ToLower(title), "生肉") {
					newVideoMsgList = append(newVideoMsgList, []string{title, url})
				}
			},
		)

		reg := regexp.MustCompile(`data_style">(.*?)</span>`)
		m := reg.FindStringSubmatch(html)
		status = m[1]
		if status == "已完结" {
			status = "已完结"
		} else {
			status = "连载中"
		}
	}

	return newVideoMsgList, status, nil
}

func yinghuacd(targetURL string) ([][]string, string, error) {
	var status string
	var newVideoMsgList [][]string

	resp, err := global.Client.R().Get(targetURL)
	if err != nil {
		global.Log.Error(err.Error())
		return newVideoMsgList, "连载中", err
	} else {
		html := goutils.EncodeToUTF8(resp)

		dom, err := goquery.NewDocumentFromReader(strings.NewReader(html))
		if err != nil {
			global.Log.Error(err.Error())
		}

		movurl := dom.Find(`div[class="movurl"]`)
		movurl.Find(`ul>li>a`).Each(func(i int, node *goquery.Selection) {
			url, _ := node.Attr("href")
			title := node.Text()
			if !strings.Contains(strings.ToLower(title), "pv") {
				newVideoMsgList = append(newVideoMsgList, []string{title, url})
			}
		})

		title := dom.Find("title").Text()
		if strings.Contains(title, "全集") {
			status = "已完结"
		} else {
			status = "连载中"
			newVideoMsgList = goutils.SliceListReverse(newVideoMsgList)
		}
	}

	return newVideoMsgList, status, nil
}

func age(targetURL string) ([][]string, string, error) {
	var status string
	var newVideoMsgList [][]string

	resp, err := global.Client.R().Get(targetURL)
	if err != nil {
		global.Log.Error(err.Error())
		return newVideoMsgList, "连载中", err
	} else {
		html := goutils.EncodeToUTF8(resp)
		dom, err := goquery.NewDocumentFromReader(strings.NewReader(html))
		if err != nil {
			global.Log.Error(err.Error())
		}

		dom.Find(`ul[class="video_detail_episode"]`).Eq(0).Find(`a`).Each(func(i int, node *goquery.Selection) {
			url, _ := node.Attr("href")
			title := node.Text()
			if !strings.Contains(strings.ToLower(title), "pv") && !strings.Contains(strings.ToLower(title), "无字") && !strings.Contains(strings.ToLower(title), "英字") && !strings.Contains(strings.ToLower(title), "生肉") {
				newVideoMsgList = append(newVideoMsgList, []string{title, url})
			}
		})

		reg := regexp.MustCompile(`detail_imform_value">(.*?)</span>`)
		m := reg.FindStringSubmatch(html)
		status = m[1]
		if status == "完结" {
			status = "已完结"
		} else {
			status = "连载中"
		}
	}

	return newVideoMsgList, status, nil
}
